### PR TITLE
Feature/1568 lv rotary

### PR DESCRIPTION
**Pull request by acumartini**
In response to requests for a rotary encoder widget #1568, I submit a `lv_rotary` implementation that extends `lv_arc`, maintains range state, and adds a `btn` knob to emulate a real digital encoder, and interface with the encoder input source..

### CODE REVIEW

**Comment by acumartini** [comment]
Fixed the tests... I had missed a bunch of minor things when I migrated my changes to my fork.

**Comment by kisvegabor** [comment]
Thank you!  

> I had missed a bunch of minor things when I migrated my changes to my fork.

That's why we have tests :slightly_smiling_face: 

I've adjusted a few parts:
- `start_value` is not required because it is related to the "left knob" of the slider. It doesn't look reasonable here
- `type` can be replaced directly with "symmetric" as there is no other mode. BTW, there was an infinite cycle in `lv_rotary_get_type`
- removed the `bg` and `indic` styles as they are already stored by the arc.
- keep the default size from the arc

**Comment by acumartini** [comment]
Awesome, thanks for the adjustments!

I had left start_value in there in case someone wanted to add starting animation back in, similar to slider functionality. I am working on the `set_value` function right now as its not quite right, then I will add the `btn` checkable functionality.

fyi: I did sign the Contributor License Agreement, not sure

**Comment by embeddedt** [comment]
The email address for your commits isn't associated with your GitHub account. That's why the CLA bot is angry. :slightly_smiling_face:

**Comment by acumartini** [comment]
@kisvegabor Q: Is adding the `btn` state to the ext the right way to add `btn` like functionality, ie `checkable` etc.? I saw that u removed it, obv because it was not being used.

**Comment by acumartini** [comment]
Also, I get this error whenever I don't cast my ext variables:
`error: invalid conversion from 'void*' to 'lv_rotary_ext_t*' [-fpermissive]`

But I notice casting is not used in other places in the codebase and I don't get errors. Very strange.

**Comment by embeddedt** [comment]
I'm 99% sure `-fpermissive` is a C++ flag; your file must be getting treated as a C++ file for some reason (even though it's named `.c`). What IDE/build system are you using?

**Comment by acumartini** [comment]
`platformio`, I think your assessment is correct.

**Comment by acumartini** [comment]
Most likely its just a missing build flag I can add in the `ini` file. I will look into it.

**Comment by acumartini** [comment]
Well anyways, we can remove the explicit casts before the final merge.

I got the value setter working, but the lv_math map functions has to be floating point... not sure if this will fly or not? Maybe there is a more math savvy way of doing this.

**Comment by embeddedt** [comment]
Core code can't rely on floating point.

**Comment by acumartini** [comment]
Can we pull the floating point into `lv_rotary`? Then people won't accidentally use it as a core functionality. Im not sure we can easily map an arbitrary range to the angle degree domain without floating point. But Im also not very good at math.

Rotary can have a define that disables range mapping, or better yet, u have to explicitly enable range mapping.

**Comment by acumartini** [comment]
What I am trying to say is that without range mapping, I think we need to pull `range` completely out of the impl. It's fine if people want to bootstrap their own range mapping on top of the rotary widget value setter to create the illusion of sensitivity.

However, this does require some intimate knowledge of how arc angles work. Not a huge deal I suppose.

**Comment by acumartini** [comment]
Alternatively, we can remove the arc/range related getters and setters and then value setting can be very straightforward for a **fixed** arc and range.

Note: People can still bootstrap mapping, its just that the arc is fixed.

**Comment by acumartini** [comment]
NVM on the `btn` Q I realize now that I understand how the indev code works that `data->state` is required to enable editablilty and it would be messy cross functionality with the rotary knob state. Having the knob be a button is attractive because it adds functionality, but obv it will have no interaction with the indev encoder data state.

**Comment by acumartini** [comment]
How does `lv_btn` override the obj ancestor signal handling for state setting to disable toggling? I have combed the code a few times now and I'm just not seeing it.

I will try to breakpoint it.

**Comment by embeddedt** [comment]
The button [drives toggling itself](https://github.com/lvgl/lvgl/blob/8dbd685a908995a951fff82a9e96b6527e0e542d/src/lv_widgets/lv_btn.c#L274-L285). Is that what you're looking for?

**Comment by acumartini** [comment]
Yeah I've been staring at that for a while. I mean the ancestor is still called before btn handles the signal itself.  Maybe it has something to do with part, `LV_BTN_PART_MAIN`. I added the same handling code to the rotary signal cb, but the knob is getting toggled regardless of `checkable` state. I will keep debugging.

**Comment by acumartini** [comment]
I pulled the checkable functionality I was working on. KISS

Alternate proposal for sensitivity, add a multiplier to ext that is used to make drag and encoder edits dramatic.  

Also, Im going to make the arc take advantage of the entire 360 degrees of implicit arc range and I am removing the range setters. Sounds like floating point is a nogo per @embeddedt so let's work with what we got.

**Comment by embeddedt** [comment]
It's possible to set a custom range on the standard slider and it doesn't use floating point. Maybe you can copy the logic from there?

**Comment by acumartini** [comment]
The issue is not setting the custom range, it's mapping the "slider" range to the arc degree domain of 0 to 360 (or if the arc range can be customized as well, anywhere from 0 to 360)

**Comment by embeddedt** [comment]
I think the slider already has to map a custom range to its actual width (e.g. if it's 20 pixels wide but the range is from 0 to 100). I could be wrong, but it looks like the same thing to me.

**Comment by acumartini** [comment]
Good call. It would have to do that. Looking.

**Comment by acumartini** [comment]
Actually the mapping is in `bar.c`. Looks simpler than what Im doing. Will explore it. However, there is the potential DivByZero error in the bar code. Prob should be a check for it: 
https://github.com/lvgl/lvgl/blob/master/src/lv_widgets/lv_bar.c#L429
https://github.com/lvgl/lvgl/blob/master/src/lv_widgets/lv_bar.c#L482

**Comment by embeddedt** [comment]
It's assumed that no one will set up a slider with 0 range (it doesn't make sense).

**Comment by acumartini** [comment]
¯\_(ツ)_/¯ Peeps are known to do some really dumb things lol

**Comment by kisvegabor** [comment]
> However, there is the potential DivByZero error in the bar code. Prob should be a check for it:

True.

Regarding mapping, my way of thinking is this: figure out the percentage of current angle relative to the min/max angle and apply this percentage to the range min/max. E.g. if the arc is at 60° in 30°.. 330° -> 10%  and the numerical range is 0..200 than the current value is 20. The rest is just playing with the order of operations to avoid rounding errors. Telling the angular range might be tricky if there is an overflow in degree. E.g. 160°..20° is 220° degree range.

**Comment by acumartini** [comment]
> just playing with the order of operations to avoid rounding errors

Yes, exactly. I will attempt to refactor the mapping function to avoid this and get back to integer math.

> 160°..20° is 220° degree range

I will need to think on this a bit more, good call out.

@kisvegabor FYI: I added a `reverse` boolean to allow the arc to be drawn from arc end to arc start, but I think it would be better to add `types` at this point to capture the three cases `NORMAL`, `SYMMETRIC`, and `REVERSE`

**Comment by acumartini** [comment]
@kisvegabor Got the `LV_ROTARY_TYPE_SYMMETRIC` impl working.

reference code:
```
    lv_obj_t *pg = lv_page_create(NULL, NULL);

    lv_obj_t *rotary = lv_rotary_create(pg, NULL);
    lv_obj_set_size(rotary, 90, 90);
    lv_obj_align(rotary, pg, LV_ALIGN_IN_TOP_LEFT, 50, 50);
    lv_rotary_set_range(rotary, -100, 100);
    lv_rotary_set_value(rotary, 100, LV_ANIM_OFF);
    lv_obj_set_event_cb(rotary, rotary_event_cb);
    lv_group_add_obj(grp_en, rotary);
    lv_rotary_set_sensitivity(rotary, 1);
    lv_rotary_set_threshold(rotary, 10);

    lv_obj_t *rotary2 = lv_rotary_create(pg, NULL);
    lv_obj_set_size(rotary2, 90, 90);
    lv_obj_align(rotary2, rotary, LV_ALIGN_OUT_RIGHT_TOP, 10, 0);
    lv_rotary_set_range(rotary2, -100, 100);
    lv_rotary_set_value(rotary2, 100, LV_ANIM_OFF);
    lv_obj_set_event_cb(rotary2, rotary_event_cb);
    lv_rotary_set_sensitivity(rotary2, 1);
    lv_rotary_set_threshold(rotary2, 10);
    lv_group_add_obj(grp_en, rotary2);
    lv_rotary_set_type(rotary2, LV_ROTARY_TYPE_SYMMETRIC);
    // lv_rotary_set_state(rotary2, LV_ROTARY_STATE_CHECKED_DISABLED);

    lv_obj_t *rotary3 = lv_rotary_create(pg, NULL);
    lv_obj_set_size(rotary3, 120, 120);
    lv_obj_align(rotary3, rotary, LV_ALIGN_OUT_BOTTOM_LEFT, 0, 10); 
    lv_rotary_set_range(rotary3, -2048, 2048);
    lv_rotary_set_value(rotary3, 1000, LV_ANIM_OFF);
    lv_obj_set_event_cb(rotary3, rotary_event_cb);
    lv_group_add_obj(grp_en, rotary3);
    lv_rotary_set_sensitivity(rotary3, 10);
    lv_rotary_set_threshold(rotary3, 50);
    lv_rotary_set_type(rotary3, LV_ROTARY_TYPE_REVERSE);  

    lv_scr_load(pg);
```

**Comment by acumartini** [comment]
Will work on fixing the mapping order of operations next.

**Comment by kisvegabor** [comment]
Great!

I've tried you test code, but can't adjust the widgets with a mouse. It did something but I couldn't set the values I wanted.

When can the "reverse" mode be useful? Can you mention a use case?

What do you think about adding these features in `lv_arc` directly? They won't bloat to widgets memory usage as there are only a few extra parameters.

**Comment by acumartini** [comment]
> It did something but I couldn't set the values I wanted.

Interesting. I've been testing with my finger on a touch screen. Likely the dragging code needs some tweaking. I added a change threshold to account for last drag x being much different the the next in a repeated swipe, but I think there is a better way.

> When can the "reverse" mode be useful? Can you mention a use case?

In my use case I want a knob to control a maximum value, and a knob to a minimum value for a parameter in program. It would be more intuitive (imo) for the max value knob indicator to be reversed to explicitly indicate to the user you are doing something very different with this knob, but to the same parameter. "Two sides same coin" kinda thing.

> What do you think about adding these features in lv_arc directly?

I'd be happy to see there where ever you think it is best. Would reduce the need for some of the getters and setters. `ARC_TYPES` could get a little confusing... but I agree its not different enough from `lv_arc` to justify a whole new widget. Maybe if it had tru button functionality? But could not figure out how to do that given there is only one EVENT_VALUE_CHANGED per object. I think the `knob` would actually need to be its own object with its own even handler. Makes things complicated enough that I punted on it. Defer to you guys on the value of that functionality.

**Comment by embeddedt** [comment]
I suspect that you are having issues controlling the slider with touch because you're just using the X coordinate to figure out where the user is touching, which doesn't work so well with a round widget. I think you need to use atan2 to figure out the relative angle of the user's touch to the center of the arc.

[This](https://github.com/lvgl/lvgl/blob/916f5b343a947ffb702adbfdbc5f165160c28554/src/lv_widgets/lv_cpicker.c#L799-L828) is the color picker's implementation; maybe that will help. If not, I can try to implement this myself at some point before this gets merged.

> I think the knob would actually need to be its own object with its own event handler.

Not necessarily; the current slider has a knob which isn't its own object.

**Comment by acumartini** [comment]
> I suspect that you are having issues controlling the slider with touch because you're just using the X coordinate to figure out where the user is touching

Right this was just to POC dragging functionality. NOTE: it only works when u drag over the `knob_area`. There is likely a million better ways to do this. Just wanted to get something down as a starting point.

**Comment by acumartini** [comment]
Fixed the rounding errors in `_lv_map`. Now all integer math.

**Comment by acumartini** [comment]
@embeddedt This color picker code is cool.

> Not necessarily; the current slider has a knob which isn't its own object.

The slider knob is not also a `btn`. Most digital encoders are also buttons, so I was thinking the UI `knob` could also have the ability to be checkable, and send its own VALUE_CHANGE events. However, in order for the knob to be a button we definitely need to move the drag tracking outside of the `knob_area` into the "ring" (arc) space.

**Comment by acumartini** [comment]
So my main Q about the `btn` functionality is, can the knob send its own VALUE_CHANGED event (when pressed or toggled) independent from the arc VALUE_CHANGED event?

If so, then I think it's worthwhile to add that functionality and move the drag sensing into the arc ring.

**Comment by kisvegabor** [comment]
> So my main Q about the btn functionality is, can the knob send its own VALUE_CHANGED event (when pressed or toggled) independent from the arc VALUE_CHANGED event?

But you can send `LV_EVENT_CLICKED` for button click and keep `LV_EVENT_VALUE_CHANGE` for moving the arc. 

Although there is no functionality in LVGL to block clicked signal manually (e.g. when the arc is changed) but it can be simply added.

**Comment by acumartini** [comment]
> But you can send LV_EVENT_CLICKED 

Got it... do you think its worth it? It's a bit outside of the scope of what I need for my project, but I want to see this widget feature complete.

**Comment by acumartini** [comment]
@embeddedt I mostly got this arc drag tracking working, however, its kind of impractical for small knobs if you are trying to use your finger on a touch screen. Probably with a mouse it would work nicely... maybe it could be optional?

```
    else if(sign == LV_SIGNAL_PRESSING) {
        lv_indev_t * indev = lv_indev_get_act();
        if(indev == NULL) return res;

        lv_indev_type_t indev_type = lv_indev_get_type(indev);
        lv_point_t p;
        if(indev_type == LV_INDEV_TYPE_ENCODER || indev_type == LV_INDEV_TYPE_KEYPAD) {
            p.x = rotary->coords.x1 + lv_obj_get_width(rotary) / 2;
            p.y = rotary->coords.y1 + lv_obj_get_height(rotary) / 2;
        }
        else {
            lv_indev_get_point(indev, &p);
        }

        if((LV_MATH_ABS(p.x - ext->last_press_point.x) > indev->driver.drag_limit / 2) ||
           (LV_MATH_ABS(p.y - ext->last_press_point.y) > indev->driver.drag_limit / 2)) {
            ext->last_press_point.x = p.x;
            ext->last_press_point.y = p.y;
        }

        /*Make point 0 relative to the rotary*/
        p.x -= rotary->coords.x1;
        p.y -= rotary->coords.y1;

        /*Ignore pressing in the inner area*/
        int16_t angle;
        lv_coord_t r_in = lv_area_get_width(&ext->knob_area) / 2;

        p.x -= r_in;
        p.y -= r_in;
        bool on_ring = true;
        if(p.x * p.x + p.y * p.y < r_in * r_in) {
            return res; /*Set the angle only if pressed on the ring*/
        }
        angle = _lv_atan2(p.x, p.y) % 360;
        
        int16_t bg_end = ext->arc.bg_angle_end;
        if (ext->arc.bg_angle_end < ext->arc.bg_angle_start) {
            bg_end = ext->arc.bg_angle_end + 360;
            if (angle < ext->arc.bg_angle_start) angle += 360;
        } 

        lv_rotary_set_value(
            rotary,
            _lv_map(angle, ext->arc.bg_angle_start, bg_end,
                    ext->min_value, ext->max_value),
            LV_ANIM_OFF
        );
    }
```

**Comment by kisvegabor** [comment]
> Got it... do you think its worth it? It's a bit outside of the scope of what I need for my project, but I want to see this widget feature complete.

We can leave it, for now, to focus on the "value set by drag".

> I mostly got this arc drag tracking working, however, its kind of impractical for small knobs if you are trying to use your finger on a touch screen. Probably with a mouse it would work nicely... maybe it could be optional?

I see that you use `lv_obj_add_protect(rotary, LV_PROTECT_PRESS_LOST);` It should allow keeping the object pressed even if you are already not pressing directly on it. So you can "turn the arc" on a larger radius, out of the arc.

**Comment by acumartini** [comment]
> I see that you use lv_obj_add_protect(rotary, LV_PROTECT_PRESS_LOST); It should allow keeping the object pressed even if you are already not pressing directly on it. So you can "turn the arc" on a larger radius, out of the arc.

Great idea. Will test. If this is still unsatisfactory I will add a drag type and setter to switch between the two modes.

**Comment by acumartini** [comment]
I got this working better by adjusting the `lv_obj_set_ext_click_area`. Not pushed yet, still tuning the value/angle setting.

**Comment by acumartini** [comment]
@kisvegabor give your mouse control another shot. I think I got this angle based drag working pretty well now.

```
    // Styles
    lv_style_init(&rotary_style);
    lv_style_set_pad_inner(&rotary_style, LV_STATE_DEFAULT, 10);

    // Home
    pg = lv_page_create(NULL, NULL);

    lv_obj_t *rotary = lv_rotary_create(pg, NULL);
    lv_obj_set_size(rotary, 90, 90);
    lv_obj_align(rotary, pg, LV_ALIGN_IN_TOP_LEFT, 50, 50);
    lv_rotary_set_range(rotary, -100, 100);
    lv_rotary_set_value(rotary, 100, LV_ANIM_OFF);
    lv_obj_set_event_cb(rotary, rotary_event_cb);
    lv_group_add_obj(grp_en, rotary);
    lv_rotary_set_sensitivity(rotary, 1);
    lv_rotary_set_threshold(rotary, 50);

    lv_obj_t *rotary2 = lv_rotary_create(pg, NULL);
    lv_obj_set_size(rotary2, 90, 90);
    lv_obj_align(rotary2, rotary, LV_ALIGN_OUT_RIGHT_TOP, 10, 0);
    lv_rotary_set_range(rotary2, -100, 100);
    lv_rotary_set_value(rotary2, 100, LV_ANIM_OFF);
    lv_obj_set_event_cb(rotary2, rotary_event_cb);
    lv_rotary_set_sensitivity(rotary2, 1);
    lv_rotary_set_threshold(rotary2, 50);
    lv_group_add_obj(grp_en, rotary2);
    lv_rotary_set_type(rotary2, LV_ROTARY_TYPE_SYMMETRIC);
    // lv_rotary_set_state(rotary2, LV_ROTARY_STATE_CHECKED_DISABLED);

    lv_obj_t *rotary3 = lv_rotary_create(pg, NULL);
    lv_obj_set_size(rotary3, 120, 120);
    lv_obj_align(rotary3, rotary, LV_ALIGN_OUT_BOTTOM_LEFT, 0, 10); 
    lv_rotary_set_range(rotary3, ROTARY_MIN, ROTARY_MAX);
    lv_rotary_set_value(rotary3, 1000, LV_ANIM_OFF);
    lv_obj_set_event_cb(rotary3, rotary_event_cb);
    lv_group_add_obj(grp_en, rotary3);
    lv_rotary_set_sensitivity(rotary3, 10);
    lv_rotary_set_threshold(rotary3, 500);
    lv_rotary_set_type(rotary3, LV_ROTARY_TYPE_REVERSE);  
    lv_obj_add_style(rotary3, LV_ROTARY_PART_KNOB, &rotary_style);

    lv_scr_load(pg);
```

**Comment by kisvegabor** [comment]
Thank you very much. I've tested and it worked really well.

I've found a conceptual thing that can improve usability. I found it easier if "dragging" keeps work once you dragged the arc. That is doesn't stop in the knob area. To start dragging you still need to press out of the dragging area. 

The rest of the changes are only minor clean up or optimizations. I've also changed the `_lv_map` function to avoid extra division. I usually use this formula and worked well so far. What do you think?

One more thing: in which cases can `threshold` be useful?

**Comment by acumartini** [comment]
This is working great! Thanks for taking a close look at this and making these improvements.

> One more thing: in which cases can threshold be useful?

I found that when dragging with a finger the value could jump around a lot because its not super precise and there could be some cases where the parameter you are adjusting is very sensitive and you want to limit the amount of change via a threshold. Setting the threshold high enough disables the limiting, which might be a good "default". Threshold setting is for special cases.

**Comment by embeddedt** [comment]
> Setting the threshold high enough disables the limiting, which might be a good "default". Threshold setting is for special cases.

I see. The standard LVGL approach to this problem is to set the range of the actual widget to be lower and then multiply the resulting value. For example, if you wanted a slider to range from 0 to 100 with a threshold of 10, the slider range should be set from 0 to 10, and the user should then multiply the return value of `lv_slider_get_value` by 10. This avoids the need to have a threshold setting directly on the widget. What do you think about this method?

**Comment by acumartini** [comment]
> For example, if you wanted a slider to range from 0 to 100 with a threshold of 10, the slider range should be set from 0 to 10, and the user should then multiply the return value of lv_slider_get_value by 10.

I don't think this prevents the value jumping that I was seeing with in-precise touch screen input. Imagine you are adjusting something like frequency of an oscillator and you want to "smooth" the adjustments out. You could put a filter on value "signal" coming out of the widget, but the arc representation on the screen would not necessarily represent what you would be hearing wrt frequency change. The threshold limiter is crude to be sure... but I do not think range mapping post value getter is quite the same thing.

I recommend we essentially turn the threshold limiting off by setting it to INT_MAX and let people know that its available in the docs if they want it.

Alternatively, if we wanted to make this more of a standard to lvgl, we could created a `misc` file with more advanced smoothing logic that can be pulled into any widget with range manipulation via input. However, Im not exactly sure what that would look like.

**Comment by kisvegabor** [comment]
@embeddedt 
I think you mixed `threshold` with  `sensitivity.


I see 2 properties which are not used by other widgets:
- `sensitivity` Tells how many steps to jump when controlled by arrows/encoder. Other widgets (e.g. slider) also has this "problem" but there the workaround is to use a range that fits well and scale the value in the event function. Regarding the resolution of the arc, setting `sensitivity = 10` or `range = 0..10` has the same effect.
- `threshold` Prevents setting too different values. No similar mechanism were used so far by other widgets. For the user experience point of you doing nothing for large moves seems a little bit counter-intuitive. Defining a slope for max change (e.g. `full rage/sec`) to make the current value closer to the goal might be better. 

What do you think?

**Comment by embeddedt** [comment]
I'm going to have to try the widget to see the difference because I'm still not getting it. :laughing:

**Comment by acumartini** [comment]
> Defining a slope for max change (e.g. full rage/sec) to make the current value closer to the goal might be better.

Agreed there are way more elegant ways to solve this and make it more intuitive. This solution would be more akin with animation and seems like a good fit. It would definitely be more user friendly as you could press anywhere on the arc and it would still respond. Also, the value increments would not need to be "step" based by some arbitrary threshold. Making it a function over time provides a lot more flexibility.

**Comment by acumartini** [comment]
@kisvegabor I think what you are proposing is actually "slew" rate limiting. I wonder if the effect I had in mind was more of a filter to smooth out noise in the touch screen input. Perhaps a better approach would be to make a filtering option in the input device code itself...

**Comment by kisvegabor** [comment]
Yes, it can be considered to a "slew rate". 

Filtering is an interesting idea. 
Due to the working mechanism of the filters, if we press the opposite side of the arc, the rate will be smaller. I'm not sure if it's the intended behavior. 

I suggest using a configurable "slew rate" in degree/sec unit. What do you think?

**Comment by acumartini** [comment]
@kisvegabor I think it is a good addition to the widget control (in place of threshold limiting). Making it tied to degrees instead of range should also make its more intuitive to set, or at least will make it more generic.

**Comment by kisvegabor** [comment]
@acumartini 
Great! Would you like to give it a try?

**Comment by acumartini** [comment]
@kisvegabor I would be happy to take a stab at this. I already have an impl of slew rate limiting handy. Do you have an example of tapping into `lvgl`'s notion of time? or should I just use timestamps to access that domain?

Sry I kinda dropped off on commits here, been focusing on pulling the widget into my project and its working awesome. But, I really want to get this feature code complete so that it can contribute to the master.

**Comment by embeddedt** [comment]
You can call `lv_tick_get` whenever you need to as a way of finding out the number of milliseconds since system startup.

**Comment by acumartini** [comment]
@kisvegabor Can you check my math here? I think I am super close to solving this but need a second pair of eyes. I will keep picking at it as well.

If I adjust this [line it](https://github.com/lvgl/lvgl/pull/1592/files#diff-0657401caf3ab2029ce83f48c0d02d25R444) begins to behave more like I would expect with the proper `threshold` setting.

ie: `int16_t delta_angle_threshold = (ext->threshold * 100) / delta_tick;`, which is obv wrong

**Comment by acumartini** [comment]
btw Im not sold on the param name `threshold`, just using it for now. Feel free to suggest a better name.

**Comment by acumartini** [comment]
I realized I had logic reversed to cancel the units out correctly, but still not behaving quite right:

```
int16_t delta_angle_threshold = (ext->threshold * delta_tick) / 1000;
```

**Comment by kisvegabor** [comment]
@acumartini 
It works well for me. I've just made a few minor fixes and refactoring but it looks good.

What issue do you see?

File: src/lv_widgets/lv_rotary.h [hunk 1]
```diff
@@ -148,7 +148,7 @@ void lv_rotary_set_sensitivity(lv_obj_t * rotary, uint16_t sensitivity);
  * @param rotary pointer to a rotary object
  * @param threshold increment threshold
  */
-void lv_rotary_set_threshold(lv_obj_t * rotary, uint16_t threshold);
+void lv_rotary_set_chg_rate(lv_obj_t * rotary, uint16_t threshold);
```
Comment (acumartini) [inline src/lv_widgets/lv_rotary.h]: 👍

File: src/lv_core/lv_obj.h [hunk 2]
```diff
@@ -102,6 +102,7 @@ enum {
     LV_EVENT_APPLY,  /**< "Ok", "Apply" or similar specific button has clicked*/
     LV_EVENT_CANCEL, /**< "Close", "Cancel" or similar specific button has clicked*/
     LV_EVENT_DELETE, /**< Object is being deleted */
+    _LV_EVENT_LAST /** Number of events*/
```
Comment (embeddedt) [inline src/lv_core/lv_obj.h:108]: Probably convenient, but never actually referenced?

Comment (acumartini) [reply src/lv_core/lv_obj.h:108]: Oh... I was added this when I was testing adding custom widget level events. Definitely should be removed given that event extension is not currently something that is needed.

**Comment by acumartini** [comment]
@kisvegabor 
So it definitely works, it's just that the `chg_rate` settings don't seem to be scaled quite right? ie, I would expect a setting of `360` to make very fast arc changes. Scaling the value up and down works as expected, such that the rate of change is very fast at 2000 and slow at 45, but that is not really what I was shooting for with degrees/second.

**Comment by acumartini** [comment]
Ah and the main issue, I forgot, is that if you set the rate lower to something like 45 it only changes in one direction... I think there is a rounding error or potentially a value caching bug.

**Comment by kisvegabor** [comment]
I've just fixed the issue with low `chg_rate` values by saving `last_tick` only when a real change occurred. So it waits until enough time elapsed to really change the value.

It introduced an asymmetry because lowering the value was slower due to rounding errors but it was fixed by adding a smal lvalue to `angle` to make it rounded in `_lv_map`

**Comment by acumartini** [comment]
Perfect, thanks for providing that sanity check. Glad my hunch as close to right.

**Comment by acumartini** [comment]
So it's moving in the positive direction now at a slower `chg_rate` but the movement seems "choppy". Are you seeing this too? Also the scale for degree/s still seems wrong, but the scaling is working.

**Comment by kisvegabor** [comment]
I saw only a minor asymmetry in  left/right direction and only with small (~45) `chg_rate`s. It didn't look choppy here.
Could you send the settings of your widget and a video about the issue?

**Comment by acumartini** [comment]
```
    lv_obj_t *rotary = lv_rotary_create(pg, NULL);
    lv_obj_set_size(rotary, 90, 90);
    lv_obj_align(rotary, pg, LV_ALIGN_IN_TOP_LEFT, 50, 50);
    lv_rotary_set_range(rotary, -100, 100);
    lv_rotary_set_value(rotary, 100, LV_ANIM_OFF);
    lv_group_add_obj(grp_en, rotary);
    lv_rotary_set_sensitivity(rotary, 1);
    lv_rotary_set_chg_rate(rotary, 45);
```
vid: https://photos.app.goo.gl/sgRFXBn3yrVLqCuP9

**Comment by acumartini** [comment]
@kisvegabor if this is only happening with a touch screen input I am fine with it. The slow rate change is a bit of an edge case anyways.

btw I tested with with an encoder input. Works really well. Overall I'm super happy with this solution. The only thing that would make it better is disabling CLICK events while dragging somehow so that the "button" functionality is more isolated.

**Comment by kisvegabor** [comment]
> Overall I'm super happy with this solution. 

I like how it works now :slightly_smiling_face: 

>  if this is only happening with a touch screen input I am fine with it. The slow rate change is a bit of an edge case anyways.

I also saw the issue by using your code. 
And I agree to leave it for now as it's really an edge case. 


> The only thing that would make it better is disabling CLICK events while dragging somehow so that the "button" functionality is more isolated.

We can't disable click event because is too low level. But we can add a function like `bool lv_rotary_is_turning(lv_obj_t * obj)`. By this the user can ignore the event as he wish.

---

Earlier we were talking about merging `lv_rotary` to `lv_arc`. What do you think about it now?

**Comment by acumartini** [comment]
> We can't disable click event because is too low level. But we can add a function like bool lv_rotary_is_turning(lv_obj_t * obj). By this the user can ignore the event as he wish.

We already have a `bool lv_rotary_is_dragged(const lv_obj_t * rotary)` (although it probably should be changed to `is_dragging` or `is_turning`), which I think fills this role. Im fine with this solution given the limitations of event hierarchies.

> Earlier we were talking about merging lv_rotary to lv_arc. What do you think about it now?

I would agree that the functionality is not different enough to justify an entirely new widget, however, it would significantly complicate the `lv_arc` code to the point where it may become hard to maintain. So my 2 cents, from a software development standpoint its cleaner to have them separate, but from a user perspective probably better to merge it in.

**Comment by kisvegabor** [comment]
> We already have a bool lv_rotary_is_dragged(const lv_obj_t * rotary) (although it probably should be changed to is_dragging or is_turning), which I think fills this role. Im fine with this solution given the limitations of event hierarchies.

Oh, I missed that.

> I would agree that the functionality is not different enough to justify an entirely new widget, however, it would significantly complicate the lv_arc code to the point where it may become hard to maintain. So my 2 cents, from a software development standpoint its cleaner to have them separate, but from a user perspective probably better to merge it in.

My reasoning is if `lv_arc` is not bloated up significantly by merging `lv_rotary` into it then it's easier to maintain one widget instead of two.

**Comment by acumartini** [comment]
> My reasoning is if lv_arc is not bloated up significantly by merging lv_rotary into it then it's easier to maintain one widget instead of two.

I fully support this, or at least giving it a shot. End of the day maintainers should make the call I think.

**Comment by acumartini** [comment]
If we move the code into `lv_arc` then the knob in the middle can be fully optional, which makes sense in a lot of cases.

**Comment by kisvegabor** [comment]
Great! :slightly_smiling_face: 

> If we move the code into lv_arc then the knob in the middle can be fully optional, which makes sense in a lot of cases

True, but it's easy to do. Just clear its styles and it will be transparent.

**Comment by acumartini** [comment]
> True, but it's easy to do. Just clear its styles and it will be transparent.

Never even thought of this. Cool! Well, you guys make the call on the port to `lv_arc`. I can prob find a bit of time to help out with it as well.

**Comment by acumartini** [comment]
@kisvegabor So where did we land on this? I'm kindov waiting for you guys to make the call here.

**Comment by kisvegabor** [comment]
> So where did we land on this? I'm kindov waiting for you guys to make the call here.

Sorry for the uncertainty. Yes, please merge it into the arc! :slightly_smiling_face:

**Comment by acumartini** [comment]
Sounds good. I will take a stab at it.

**Comment by acumartini** [comment]
Sorry I have been AFK on this for a bit. Work has been very stressful and haven't had a ton of energy outside of that to code. I plan to have a window to focus on this and wrap it up in about 2 weeks.

**Comment by kisvegabor** [comment]
No problem @acumartini. I hope the stressful part of your work will be over soon. Hang in there! :slightly_smiling_face: 

This feature will be required in one of my projects this week. So I can also do it if you are busy.

**Comment by kisvegabor** [comment]
I merge rotary to arc. 
It looks quite the same as before. The only change I made is the removal of `sensitivity` because the very same effect can be achieved by settings a smaller range and scaling the values in the event function.

Could you test it if you have some time?

**Comment by kisvegabor** [comment]
Finally, I've changed one more thing: the idea of "knob". In slider and switch, "knob" means the circle which indicates the current position. I found it it make more sense here too. 

The original idea of "knob" can be easily mimicked by placing a button on the arc, however, the current version would be very difficult for a user to implement. 

It looks like this:
image

Paddings can be used to make the knob larger or smaller.

Let me know what you think!

**Comment by kisvegabor** [comment]
I was testing this feature in the past days and it worked well for me. (Not tested every feature though).

I merge it into the `dev` branch so it it will be part of the next release. 

@acumartini 
Thank you very much for your work on this update!

**Comment by embeddedt** [comment]
@kisvegabor I think there should be an API that lets you disable the knob on an arc. I don't see any way to do this at the moment besides making the knob transparent and disabling click events on the arc object.

**Comment by kisvegabor** [comment]
It's really a little bit cumbersome now.
What about  `lv_arc_set_adjustabe()`?

**Comment by embeddedt** [comment]
That would work fine.

**Comment by acumartini** [comment]
@kisvegabor @embeddedt Thanks for the collaboration and pulling this forward into your dev branch! Can't wait for the next release. Sorry for not responding, I had to back burner this for a bit. Things have finally settled down at work and preparing for a nice vacation :) I will be testing this code on my personal project now. Happy to help with improvements in the future.

**Comment by acumartini** [comment]
Already been release. Nice! You guys move fast. Merge into lv_arc looks really good. Feature is working perfectly. Knob on the indicator is way more intuitive. Great call and good work. 🎸

**Comment by kisvegabor** [comment]
Great! Thank you for the feedback!

**Comment by acumartini** [comment]
FYI: This slipped through the cracks in the conf_template: https://github.com/lvgl/lvgl/blob/master/lv_conf_template.h#L690

**Comment by acumartini** [comment]
@kisvegabor
Heads up I found a bug on the master branch. Haven't discovered the root cause yet, but try this out:

```
    lv_obj_t *rotary = lv_arc_create(pg, NULL);
    lv_obj_set_size(rotary, 90, 90);
    lv_obj_align(rotary, pg, LV_ALIGN_IN_TOP_LEFT, 50, 50);
    lv_arc_set_range(rotary, -20000, 20000);
    lv_arc_set_value(rotary, 0);
    lv_obj_set_event_cb(rotary, rotary_event_cb);
    lv_group_add_obj(grp_en, rotary);
    lv_arc_set_chg_rate(rotary, 500);
```

You would expect the Indicator knob to be at the top of the arc in the middle, but its actually at the MIN value.  I'm guessing it is a 0 value error, but I am not sure where/how yet.

**Comment by acumartini** [comment]
Found the issue.
https://github.com/lvgl/lvgl/blob/master/src/lv_widgets/lv_arc.c#L389
https://github.com/lvgl/lvgl/blob/master/src/lv_widgets/lv_arc.c#L395

The early returns make no sense if there was a range change. I think we need to remove them or add a boolean to disable return checks for the next call to the `set_value` function in the event of a range change.

**Comment by acumartini** [comment]
@kisvegabor @embeddedt 
Moving this to a new PR with my proposed patch: https://github.com/lvgl/lvgl/pull/1747

... just realized this is a PR to the wrong branch. Fixing

**Comment by acumartini** [comment]
OK this looks right: https://github.com/lvgl/lvgl/pull/1748

**Pull request merged**